### ISSUE TITLE
explore view functions as onchain logic for offchain backends with hidden information

### ISSUE THREAD

**Issue by holic** [latticexyz/mud#3127]
Let's say I have a game that's hybrid onchain/offchain where I want to enable players to control in-game entities with onchain logic. Let's also say that player position is only offchain and not meant to be broadcasted (otherwise you might get killed or robbed or whatever).

Using Minecraft as an example, I want to enable players to have a "programmable door" where the player can control whether the door will open/close when interacted with. The open/close state is controlled by the offchain backend:
- the game client sends a request to the backend to open the door
- the backend asks the programmed door's view function if the door should be opened
- if so, the backend changes the door state to open

Let's say we only want to allow Alice to open the door if Bob is nearby.

```solidity
function canOpenDoor() public view returns (bool) {
  Position alice = getPlayerPosition(_msgSender());
  Position bob = getPlayerPosition("0xB0B");
  return getDistance(alice, bob) < 4;
}
```

Since we don't want to reveal positions onchain, the function `getPlayerPosition` would use a strategy similar to [CCIP-Read](https://eips.ethereum.org/EIPS/eip-3668) where it leaves a placeholder (a revert, in the CCP-Read case) that the backend can fill in when it computes the return value of the view function.

**Comment by qbzzt** [comment]
My main problem here is that while we don't reveal the locations to the player, we do reveal it to player-written code. Imagine a door that only opens when `location.x %2 = 1`, another that only opens when `location.x % 4 >= 2`, a third with `location.x % 8 >= 4`, etc. As much as generic solutions are preferable in general, I think here we'll need have the backend provide a function such as `distanceBetweenPlayers`.



Let me see if I understand the process.

1. The client sends a request to the backend for something (for example, `openDoor`).
2. The backend sends a request to a node to run a view function (in this case, `canOpenDoor`) written as part of an extension (which is why it's not directly in the backend).
3. The node reverts, and sends the revert message to the backend
4. The backend provides alice's position and calls the view function again.
5. The node reverts again, asking for bob's position
6. The backend provides bob's position and calls the view function a third time.
7.  The node calculates the distance, and responds with a `true` or `false`.
8. The backend opens the door (or not) and responds to the client.

This requires us to trust the node not to abuse the location information, which isn't a terrible requirement (we can ask the game develop to run their own node). It's a complex process, which will probably be time consuming. Maybe we can do better with tevm.

**Comment by holic** [comment]
> we do reveal it to player-written code

what is the concern here? if it's a view function, there's no way for the view function to write to anything or communicate externally to potentially expose the data

**Comment by qbzzt** [comment]
> if it's a view function, there's no way for the view function to write to anything or communicate externally to potentially expose the data

But there is. The `view` function can only return a value to the backend. However, that value visibly affects gameplay. This gives us a side channel to smuggle out information.

For example, imagine that the game includes a door that can be open or closed depending on Alice-provided code. A different player, Bob, can be in one of 64 different locations (an 8x8 map).

Alice will create six doors, and link them to six functions.

```
door0 <- Bob.position.x % 2 == 1
door1 <- Bob.position.x % 4 >= 2
door2 <- Bob.position.x % 8 >= 4
door3 <- Bob.position.y % 2 == 1
door4 <- Bob.position.y % 4 >= 2
door5 <- Bob.position.y % 8 >= 4
```

If we know that only `door0`, `door2`, and `door5` are open, then we know that Bob is at position `(5,4)`.

**Comment by roninjin10** [comment]
This reminds me of the Hevm style symbolic execution research I'm doing for Tevm

**Comment by holic** [comment]
> > if it's a view function, there's no way for the view function to write to anything or communicate externally to potentially expose the data
> 
> But there is. The `view` function can only return a value to the backend. However, that value visibly affects gameplay. This gives us a side channel to smuggle out information.
> 
> For example, imagine that the game includes a door that can be open or closed depending on Alice-provided code. A different player, Bob, can be in one of 64 different locations (an 8x8 map).
> 
> Alice will create six doors, and link them to six functions.
> 
> ```
> door0 <- Bob.position.x % 2 == 1
> door1 <- Bob.position.x % 4 >= 2
> door2 <- Bob.position.x % 8 >= 4
> door3 <- Bob.position.y % 2 == 1
> door4 <- Bob.position.y % 4 >= 2
> door5 <- Bob.position.y % 8 >= 4
> ```
> 
> If we know that only `door0`, `door2`, and `door5` are open, then we know that Bob is at position `(5,4)`.

I guess what you're hinting at is a kind of like a timing attack? Bob would have to interact with all the doors and Alice would need to know that Bob was the last to interact with all the doors?

**Comment by qbzzt** [comment]
> I guess what you're hinting at is a kind of like a timing attack? To get to that answer above, Bob would have to interact with all the doors and Alice would need to know that Bob was the last to interact with all the doors?

There is no need for timing or for Bob to do anything. Alice interacts with the doors to get the backend to call the `view` functions; it's just that those functions look at Bob's position. Bob's identity can be hard-wired in the contract or specified by other means. Theoretically, the contract can have a "setVictim" function that Alice uses to specify which player's location she wants.

**Comment by holic** [comment]
Ohh I see what you mean now: Alice could program her thing with arbitrary logic that can slowly reveal hidden information.

**Comment by qbzzt** [comment]
Yes. It's a clever solution that would be great,  except that any time I can write code that can affect something I can observe I can smuggle information out.
